Add support to filter nodes by node flags

jonasschnelli opened the pull request:
needs testing... 
needs review...

sipa commented on main.cpp:
```diff
@@ -122,6 +125,18 @@ class CDnsSeedOpts {
 
+        case 's': {
+          if (!strcmp(optarg, "NODE_NETWORK"))
+            requiredNodeFlags = NODE_NETWORK;
```
Maybe use `|=` instead of `=`, so that this flag can be specified multiple times?

jonasschnelli replied:
Yes. Good idea.
But how would you parse the `optarg`?
We could use a one char flag identifier (B for BLOOM, W for WITNESS) and use a optarg of `BW`?

jonasschnelli commented:
Updated:

The response depends now on the requesting subdomain.
Example:
hostname: seed.bitcoin.jonasschnelli.ch
* w.seed.bitcoin.jonasschnelli.ch -> results only nodes with enables `NODE_WITNESS` service bit
* b.seed.bitcoin.jonasschnelli.ch -> results only nodes with enables `NODE_BLOOM` service bit
* g.seed.bitcoin.jonasschnelli.ch -> results only nodes with enables `NODE_GETUTXO` service bit
* bg.seed.bitcoin.jonasschnelli.ch -> results only nodes with enables `NODE_GETUTXO & NODE_BLOOM` service bit
* ...

jonasschnelli commented:
Running now on: testnet-seed.bitcoin.jonasschnelli.ch (or w.testnet-seed.bitcoin.jonasschnelli.ch, gb.testnet-seed.bitcoin.jonasschnelli.ch)

jonasschnelli commented:
Updated.
Removed string->bitflag mapping.
Now the seeder accepts filtering by `<hex uint64_t>.seeder.domain.tld`.

Example:
`dig 5.testnet-seed.bitcoin.jonasschnelli.ch` -> will return nodes with NODE_NETWORK | NODE_BLOOM.
`dig 9.testnet-seed.bitcoin.jonasschnelli.ch` -> will return nodes with NODE_NETWORK | NODE_WITNESS.

jonasschnelli commented:
Updated.
The hostname now needs to start with `x`.
Example: `x5.testnet-seed.bitcoin.jonasschnelli.ch` for nodes with `NODE_BLOOM`.

Also the seeder will reject (use default `NODE_NETWORK` service bits filter) if there are more then 16 hex digits.

sipa commented on main.cpp:
```diff
@@ -227,24 +227,47 @@ class CDnsThread {
+  int hostlen = strlen(requestedHostname);
+  char maxHex[16+1]; //uint64_t as hex + null terminator
+  if (hostlen > 0 && requestedHostname[0] == 'x') {
+    for (unsigned int i = 1; i<hostlen; i++)
```
I think this whole loop can be replaced by a call to strtoul with endptr passed in, and then checking whether at return endptr points to a '.'.

jonasschnelli replied:
* How would you then detect subdomains longer then 17 chars?
* How would you then reject non hex chars?
* How would you then reject leading zeros?

sipa replied:
> How would you then detect subdomains longer then 17 chars?

if (endptr > 17+ptr) {

> How would you then reject non hex chars?

strtoul stops at the first non-hex character. Just check whether endptr points to a '.'.

> How would you then reject leading zeros?

Check that the first character after 'x' is not '0'.

jonasschnelli replied:
Okay. Sounds good and makes sense. Will change soon.

jonasschnelli commented:
@sipa: simplified the `strtoull` logic (removed the loop).

sipa commented on db.cpp:
```diff
@@ -175,22 +175,31 @@ void CAddrDb::GetIPs_(set<CNetAddr>& ips, int max, const bool* nets) {
       id = *ourId.begin();
     }
-    if (id >= 0) {
+    if (id >= 0 && idToInfo[id].services & requestedFlags) {
```
Does this not need a == requestedFlags?

jonasschnelli replied:
IMO this is correct. 
` && idToInfo[id].services & requestedFlags` should filter all services that do not have the requestedFlags set (not == but bit AND check). Or am I wrong?

sipa replied:
No, I think it needs to be:
```
if (id >= 0 && (idToInfo[id].services & requestedFlags) == requestedFlags) {
```

Otherwise if you ask for flags 5, you may get entries with flags 1 or 4.

sipa commented on db.cpp:
```diff
@@ -175,22 +175,31 @@ void CAddrDb::GetIPs_(set<CNetAddr>& ips, int max, const bool* nets) {
-  if (max > goodId.size() / 2)
-    max = goodId.size() / 2;
+  std::set<int> goodIdFiltered;
+  for (std::set<int>::const_iterator it = goodId.begin(); it != goodId.end(); it++) {
```
Using a vector instead of a set would be more efficient, as we already know they're all different anyway.

The whole logic below that uses range/low/high can go away then, and instead pick `goodIdFiltered[rand() % goodIdFiltered.size()]`.

jonasschnelli commented:
@sipa: switched from std::set to std::vector for IP query mechanism.

sipa commented:
I think we'll need a whitelist of acceptable flags. Currently I can iterate through 2^64 combinations, causing you to create 2^64 caches...

sipa commented on main.cpp:
```diff
@@ -167,48 +167,48 @@ extern "C" void* ThreadCrawler(void* data) {
-  vector<addr_t> cache;
+  std::map<uint64_t, vector<addr_t> > cache;
   int nIPv4, nIPv6;
   time_t cacheTime;
```
I think cacheTime should also be indexed by service bits.

jonasschnelli commented:
Good point! I'll add a whitelist.
I think we should also reject non-valid requests (reject = empty response). Things like "foo.seed.tld" where "seed.tld" is the valid domain.

jonasschnelli commented:
* Added a whitelist for possible bitmaps.
* Reject (empty response) unknown query subdomains (not matching the whitelist service filter)
* Added a std::map for cacheTime

NOX-PACT reviewed (commented):
Ad no regresthe for Grajnd Mendatairy of lnternational of Apocalypse of The NoX-PacT of NAuthority and Authority of districk power gestioning all the world of Legion of the districk felipine the sovit subvensionalyse this signed general of K.g.b kevin pare
